src/plugins/csp_handler/server/csp_handlers.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { ConfigurationClient } from '../../application_config/server';
+import {
+  CoreSetup,
+  IScopedClusterClient,
+  Logger,
+  OnPreResponseHandler,
+  OnPreResponseInfo,
+  OnPreResponseToolkit,
+  OpenSearchDashboardsRequest,
+} from '../../../core/server';
+
+const CSP_RULES_CONFIG_KEY = 'csp.rules';
+
+/**
+ * This function creates a pre-response handler to dynamically set the CSP rules.
+ * It give precedence to the rules from application config plugin over those from YML.
+ * In case no value from application config, it will ensure a default frame-ancestors is set.
+ *
+ * @param core Context passed to the plugins `setup` method
+ * @param cspHeader The CSP header from YML
+ * @param getConfigurationClient The function provided by application config plugin to retrieve configurations
+ * @param logger The logger
+ * @returns The pre-response handler
+ */
+export function createCspRulesPreResponseHandler(
+  core: CoreSetup,
+  cspHeader: string,
+  getConfigurationClient: (scopedClusterClient: IScopedClusterClient) => ConfigurationClient,
+  logger: Logger
+): OnPreResponseHandler {
+  return async (
+    request: OpenSearchDashboardsRequest,
+    response: OnPreResponseInfo,
+    toolkit: OnPreResponseToolkit
+  ) => {
+    try {
+      const shouldCheckDest = ['document', 'frame', 'iframe', 'embed', 'object'];
+
+      const currentDest = request.headers['sec-fetch-dest'];
+
+      if (!shouldCheckDest.includes(currentDest)) {
+        return toolkit.next({});
+      }
+
+      const [coreStart] = await core.getStartServices();
+
+      const myClient = getConfigurationClient(coreStart.opensearch.client.asScoped(request));
```

Comment:
can we have a better name like configClient instead of myClient?

Reply:
+1

Reply:
updated